src/main/java/fi/metatavu/metamind/server/keycloak/AuthenticationController.java

```diff
@@ -0,0 +1,408 @@
+        policyRepresentation.setName("user-" + userName);
+        policyRepresentation.setDecisionStrategy(DecisionStrategy.AFFIRMATIVE);
+        policyRepresentation.addUser(userName);
+        policyRepresentation.setId(userId.toString());
+        userPolicies.create(policyRepresentation);
+
+      }
+    }
+     return result; 
+    }
+    
+
+  
+  /**
+   * Returns list of permitted users for a resource with given scopes 
+   * 
+   * @param keycloak keycloak client instance
+   * @param realmName realm name
+   * @param clientId client id
+   * @param resourceId resource id
+   * @param resourceName resource name
+   * @param scopes scopes
+   * @return set of user ids
+   */
+  public Set<UUID> getResourcePermittedUsers(String realmName, UUID resourceId, String resourceName, List<AuthorizationScope> scopes) {
+    Keycloak keycloak = getAdminClient();
+    ClientRepresentation client = getClient(keycloak);
+    RealmResource realm = keycloak.realm(realmName);
+    return getPermittedUsers(realm, client, resourceId, resourceName, scopes);      
+  }
+  
+  /**
+   * Finds a id from Keycloak create response 
+   * 
+   * @param response response object
+   * @return
+   */
+  public static UUID getCreateResponseId(Response response) {
+    if (response.getStatus() != 201) {
+      try {
+        if (logger.isErrorEnabled()) {
+          logger.error("Failed to execute create: {}", IOUtils.toString((InputStream) response.getEntity(), StandardCharsets.UTF_8));
+        }
+      } catch (IOException e) {
+        logger.error("Failed to extract error message", e);
+      }
+      
+      return null;
+    }
+   
+    String location = response.getHeaderString("location");
+    if (StringUtils.isBlank(location)) {
+      return null;
+    }
+    
+    Pattern pattern = Pattern.compile(".*\\/(.*)$");
+    Matcher matcher = pattern.matcher(location);
+    
+    if (matcher.find()) {
+      return UUID.fromString(matcher.group(1));
+    }
+    
+    return null;
+  }
+  
+  private static Set<UUID> getPermittedUsers(RealmResource realm, ClientRepresentation client, UUID resourceId, String resourceName, List<AuthorizationScope> scopes) {
+    Map<UUID, DecisionEffect> policies = evaluatePolicies(realm, client, resourceId, resourceName, scopes);
+    
+    return policies.entrySet().stream()
+        .filter(entry -> DecisionEffect.PERMIT.equals(entry.getValue()))
+        .map(Entry::getKey)
+        .collect(Collectors.toSet());
+  }
+  
+  /**
+   * Evaluates policies resource for realm users 
+   * 
+   * @param realm realm name
+   * @param client client
+   * @param resourceId resource id
+   * @param resourceName resource name
+   * @param scopes scopes
+   * @return map of results where key is user id and value is decision
+   */
+  private static Map<UUID, DecisionEffect> evaluatePolicies(RealmResource realm, ClientRepresentation client, UUID resourceId, String resourceName, List<AuthorizationScope> scopes) {
```

Comment:
Permissions should be evaluated with policy enforcers instead of code, is this a debug method?